src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -260,90 +266,146 @@ bool GcsClient::GetGcsServerAddressFromRedis(redisContext *context,
-    ReconnectGcsServer();
+    accepted = ReconnectGcsServerAsync(failure_handled_callback);
     break;
   case rpc::GcsServiceFailureType::GCS_SERVER_RESTART:
     // If GCS sever address has changed, reconnect to GCS server and redo
     // subscription.
-    ReconnectGcsServer();
-    // If using GCS server for pubsub, resubscribe to GCS publishers.
-    resubscribe_func_(RayConfig::instance().gcs_grpc_based_pubsub());
-    // Resend resource usage after reconnected, needed by resource view in GCS.
-    node_resource_accessor_->AsyncReReportResourceUsage();
+    accepted = ReconnectGcsServerAsync([this]() {
+      // If using GCS server for pubsub, resubscribe to GCS publishers.
+      resubscribe_func_(RayConfig::instance().gcs_grpc_based_pubsub());
+      // Resend resource usage after reconnected, needed by resource view in GCS.
+      node_resource_accessor_->AsyncReReportResourceUsage();
+    });
     break;
   default:
     RAY_LOG(FATAL) << "Unsupported failure type: " << type;
     break;
   }
+
+  return accepted;
 }
 
-void GcsClient::ReconnectGcsServer() {
+bool GcsClient::ReconnectGcsServerAsync(
+    const std::function<void()> reconnected_callback) {
+  absl::MutexLock l(&reconnect_flag_mutex_);
+  if (disconnected_) {
+    return false;
+  }
+  {
+    absl::MutexLock l(&reconnect_callbacks_mutex_);
+    reconnected_callbacks_.emplace_back(reconnected_callback);
+  }
+  if (reconnect_in_progress_) {
+    RAY_LOG(DEBUG) << "GcsClient reconnect in progress, the callback is pended.";
+    return false;
+  }
+  reconnect_in_progress_ = true;
+  DoReconnect(absl::Now());
+  return true;
+}
+
+void GcsClient::DoReconnect(absl::Time start) {
+  if (disconnected_) {
+    ExecuteReconnectedCallbacks();
+    reconnect_in_progress_ = false;
+    return;
+  }
   std::pair<std::string, int> address;
-  auto timeout_s =
-      absl::Seconds(RayConfig::instance().gcs_rpc_server_reconnect_timeout_s());
-  auto start = absl::Now();
-  auto reconnected = false;
-  while (absl::Now() - start < timeout_s) {
-    if (disconnected_) {
+  if (get_server_address_func_(&address)) {
+    // After GCS is restarted, the gcs client will reestablish the connection. At
+    // present, every failed RPC request will trigger `ReconnectGcsServerAsync`. In order
+    // to avoid repeated connections in a short period of time, we add a protection
+    // mechanism: if the address does not change (meaning gcs server doesn't restart),
+    // the connection can be made at most once in
+    // `minimum_gcs_reconnect_interval_milliseconds` milliseconds.
+    if (last_reconnect_address_ == address &&
+        (current_sys_time_ms() - last_reconnect_timestamp_ms_) <
+            RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()) {
+      RAY_LOG(DEBUG)
+          << "Repeated reconnection in "
+          << RayConfig::instance().minimum_gcs_reconnect_interval_milliseconds()
+          << " milliseconds, return directly.";
```

Comment:
This part seems wrong. The comment says this should be no-op when there are many reconnection requests because it adds additional pings to GCS. With the new logic, it won't happen because `reconnect_in_progress_  == true`. Maybe we should just remove this logic?

Reply:
It is legacy code, maybe can be removed since we have `reconnect_in_progress_ ` now.